packages/build-utils/src/get-project-paths.ts

```diff
@@ -0,0 +1,63 @@
+import { detectFramework } from './detect-framework';
+import { DetectorFilesystem } from './detectors/filesystem';
+import frameworks from '@vercel/frameworks';
+
+const MAX_DEPTH_TRAVERSE = 3;
+
+export interface GetProjectPathsOptions {
+  fs: DetectorFilesystem;
+  path?: string;
+  skipPaths?: string[];
+}
+
+export type ProjectPath = string;
+
+export const getProjectPaths = async ({
+  fs,
+  path,
+  skipPaths,
+}: GetProjectPathsOptions): Promise<ProjectPath[]> => {
+  const allPaths: Array<ProjectPath> = [];
+  for (
+    let currentDepth = 0;
+    currentDepth < MAX_DEPTH_TRAVERSE;
+    currentDepth++
+  ) {
+    if (path && skipPaths?.includes(path)) {
+      return allPaths;
+    }
+    const directoryContents = await fs.readdir(path ?? './');
```

Comment:
At the leaf level this `readdir` call is unnecessary; we only need to know if a folder contains `package.json` and we can check that location directly. Effectively we can check N+1 levels deep while only traversing N. This could matter a lot for performance in some repositories.

Reply:
I am understanding half of this. But how can you check a location directly for `package.json` of child directories without getting the contents of the directory you are in?

Reply:
If i already have all the directories then I can do:
But I would still need to go through for the children directories unless those are already give too

How do I get all the paths?
`directoryPaths.forEach(path => {
  const hasPackageJson = fs.isFile(path);
  if (hasPackageJson) {
    allPaths.push(path)
  }
})`

Reply:
I updated the function where it matches the structure not excessively calling `readdir`